bee-message/src/message.rs

```diff
@@ -171,64 +176,67 @@ impl Message {
-            + self.parents.packed_len()
-            + option_payload_packed_len(self.payload.as_ref())
-            + self.nonce.packed_len()
+    /// Consumes the [[`Message`]], and returns ownership over its [`Parents`] and [`Payload`].
+    pub fn into_parents_and_payload(self) -> (Parents, Option<Payload>) {
+        (self.parents, self.payload.into())
     }
+}
 
-    fn pack<W: Write>(&self, writer: &mut W) -> Result<(), Self::Error> {
-        self.network_id.pack(writer)?;
-        self.parents.pack(writer)?;
-        option_payload_pack(writer, self.payload.as_ref())?;
-        self.nonce.pack(writer)?;
+impl bee_packable::Packable for Message {
+    type UnpackError = Error;
+
+    fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error> {
+        self.network_id.pack(packer)?;
+        self.parents.pack(packer)?;
+        self.payload.pack(packer)?;
+        self.nonce.pack(packer)?;
 
         Ok(())
     }
 
-    fn unpack_inner<R: Read + ?Sized, const CHECK: bool>(reader: &mut R) -> Result<Self, Self::Error> {
-        let network_id = u64::unpack_inner::<R, CHECK>(reader)?;
+    fn unpack<U: Unpacker, const VERIFY: bool>(
+        unpacker: &mut U,
+    ) -> Result<Self, UnpackError<Self::UnpackError, U::Error>> {
+        let network_id = u64::unpack::<_, VERIFY>(unpacker).infallible()?;
 
-        let parents = Parents::unpack_inner::<R, CHECK>(reader)?;
+        let parents = Parents::unpack::<_, VERIFY>(unpacker)?;
 
-        let (payload_len, payload) = option_payload_unpack::<R, CHECK>(reader)?;
+        let payload = OptionalPayload::unpack::<_, VERIFY>(unpacker)?;
 
-        if CHECK
+        if VERIFY
             && !matches!(
-                payload,
+                *payload,
                 None | Some(Payload::Transaction(_)) | Some(Payload::Milestone(_)) | Some(Payload::Indexation(_))
             )
         {
             // Safe to unwrap since it's known not to be None.
-            return Err(Error::InvalidPayloadKind(payload.unwrap().kind()));
+            return Err(UnpackError::Packable(Error::InvalidPayloadKind(
+                Into::<Option<Payload>>::into(payload).unwrap().kind(),
+            )));
         }
 
-        let nonce = u64::unpack_inner::<R, CHECK>(reader)?;
+        let nonce = u64::unpack::<_, VERIFY>(unpacker).infallible()?;
 
-        // Computed instead of calling `packed_len` on Self because `payload_len` is already known and it may be
-        // expensive to call `payload.packed_len()` twice.
-        let message_len = network_id.packed_len() + parents.packed_len() + payload_len + nonce.packed_len();
+        let message = Self {
+            network_id,
+            parents,
+            payload,
+            nonce,
+        };
 
-        if CHECK && message_len > Message::LENGTH_MAX {
-            return Err(Error::InvalidMessageLength(message_len));
+        let message_len = message.packed_len();
+
+        // FIXME: compute this in a more efficient way.
```

Comment:
yeah, I'm not sure how to fix this quickly. What would you and @thibault-martinez say about changing the return type of `Packable::unpack` so it returns `(Self, usize)` instead of just `Self`? the `usize` would be the number of read bytes.

Reply:
Let's leave it like that for now and maybe open an issue to further discuss it ?